### PR TITLE
feat: Emg pipeline improvement

### LINKED ISSUE #21: EMG 실데이터 기반 signal processing 및 inference pipeline 개선
## Title
EMG 실데이터 기반 signal processing 및 inference pipeline 개선

---

## Description
실제 EMG 데이터 수신을 기반으로 전체 파이프라인(stream → signal processing → feature extraction → inference)을 점검하고 개선한다.  
노이즈, 비정상 데이터, 사용자별 차이를 고려하여 전처리 및 추론 정확도를 향상시키고, calibration 결과를 반영한 개인화된 inference 구조를 구축한다.

---

## Scope
- EMG 실시간 데이터 수신 안정화
- signal preprocessing 개선
- feature extraction 고도화
- inference 모델 개선
- calibration 기반 개인화 적용
- end-to-end pipeline 통합 및 검증

---

### 1. EMG 데이터 수신 및 검증
- [ ] EMG window validation (NaN, empty, outlier)
- [ ] sequenceNumber 누락 및 역전 처리
- [ ] samplingRate / windowSize 일관성 검증
- [ ] session / calibration 상태 기반 수신 제한
- [ ] EmgWindowAck 실패 응답 명확화

---

### 2. Signal Processing 개선
- [ ] band-pass filter 적용 (노이즈 제거)
- [ ] rectification (절댓값 처리)
- [ ] smoothing (moving average / low-pass)
- [ ] baseline normalization (calibration 기반)
- [ ] fatigue 대응 scaling 로직 추가

---

### 3. Feature Extraction 개선
- [ ] MAV, RMS 재검증 (실데이터 기준)
- [ ] Zero Crossing (ZC) 추가
- [ ] Slope Sign Change (SSC) 추가
- [ ] Waveform Length (WL) 추가
- [ ] feature vector normalization

---

### 4. Inference 모델 개선
- [ ] 기존 모델 구조 점검 (intent_model.pkl)
- [ ] rule-based threshold + ML hybrid 구조 적용
- [ ] STOP 채널 우선 처리 로직 추가
- [ ] confidence score 반환 구조 추가
- [ ] inference latency 최적화

---

### 5. Calibration 연동
- [ ] calibration 결과 → inference 전달 구조 정의
- [ ] activationThreshold 적용
- [ ] intentThresholds 적용
- [ ] 사용자별 baseline 반영
- [ ] calibration 미존재 시 fallback 처리

---

### 6. End-to-End Pipeline 통합
- [ ] stream → processing → feature → inference 연결
- [ ] window 단위 실시간 추론 구현
- [ ] 결과 API/WebSocket 반환
- [ ] latency 측정 및 병목 분석
- [ ] 안정성 테스트 (실데이터 기반)

---

## Acceptance Criteria
- 실데이터 입력 시 pipeline이 중단 없이 동작
- 비정상 데이터가 들어와도 시스템 안정성 유지
- calibration 반영된 inference 결과 출력
- 평균 inference latency 기준 만족 (목표 정의 필요)

### PR DESCRIPTION
**Pull request by onowoey**
## PR 개요
EMG 실데이터 기반 signal processing 및 inference pipeline 개선 작업입니다.  
신호 전처리 방식, threshold 계산 로직, calibration 분기 처리 등을 전반적으로 수정하였습니다.

---

### Signal Processing 개선
- Feature extraction 이전에 신호 전처리 수행
  - DC offset 제거
  - Rectification(정류) 적용
- 기존 정규화(normalization) 제거
  - amplitude 정보 보존 목적

### Feature / Threshold 로직 개선
- signal processing → feature extraction → threshold 계산 순서로 pipeline 재구성
- calibration 유무에 따라 threshold 계산 방식 분기 처리
  - calibration 존재 시: 비율 기반 threshold
  - calibration 미존재 시: CV(Coefficient of Variation) 기반 threshold

### Architecture 개선
- singleton instance 구조 추가
- inference 관련 instance 관리 로직 정리

### 테스트 및 검증
- mock EMG WebSocket client 추가
- 실시간 inference 테스트 환경 구성

---

## 기대 효과
- 실데이터 기반 inference 안정성 향상
- 노이즈 및 DC bias 영향 감소
- feature consistency 개선
- calibration 유무에 따른 adaptive threshold 적용 가능
- 실시간 EMG streaming 테스트 용이성 향상

## Summary by CodeRabbit

* **New Features**
  * Added a WebSocket-based mock EMG data sender with demo scenarios, per-window sample messages, and a CLI flag to control failsafe edge-case tests.

* **Improvements**
  * More accurate calibration results using signal-processing computations.
  * Refined signal quality and fatigue scoring for clearer, more stable metrics.
  * Training now aggregates multiple data files and applies baseline-aligned preprocessing.
  * Backend intent-sending now returns explicit success status.
  * Signal preprocessing defaults no longer auto-normalize unless enabled.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Capstone-EMG-3P1L/NeuroMove_AI/pull/23?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: training/train_intent_model.py [hunk 1]
```diff
@@ -95,6 +120,24 @@ def _map_labels(labels: pd.Series) -> pd.Series:
...
+    processed = np.zeros_like(samples)
+
+    for i in range(samples.shape[0]):
+        signal = samples[i]
+        # DC 제거하지 않음
+        # signal = remove_dc_offset(signal)
+        signal = rectify_signal(signal)
+        processed[i] = signal
+
+    return processed
```
[the remark this answers is not in the document]
Comment (onowoey) [reply training/train_intent_model.py]: Verified against the current code.

Skipped both findings because they are no longer valid.

- The preprocessing pipeline no longer uses the old `_preprocess_samples()` flow.
- Training preprocessing now computes a REST baseline per channel and applies z-score normalization before rectification via `_apply_zscore_rectify()`.
- The comment is also consistent with the current implementation: REST baseline 기반 z-score + 정류.

No additional code changes were needed.
